packages/file_selector/file_selector_macos/test/file_selector_macos_test.dart

```diff
@@ -344,6 +344,126 @@ void main() {
+      expect(options.nameFieldStringValue, null);
+      expect(options.prompt, null);
+    });
+
+    test('handles cancel', () async {
+      when(mockApi.displaySavePanel(any)).thenAnswer((_) async => null);
+
+      final FileSaveLocation? location = await plugin.getSaveLocation();
+
+      expect(location, null);
+    });
+
+    test('passes the accepted type groups correctly', () async {
+      const XTypeGroup group = XTypeGroup(
+        label: 'text',
+        extensions: <String>['txt'],
+        mimeTypes: <String>['text/plain'],
+        uniformTypeIdentifiers: <String>['public.text'],
+      );
+
+      const XTypeGroup groupTwo = XTypeGroup(
+          label: 'image',
+          extensions: <String>['jpg'],
+          mimeTypes: <String>['image/jpg'],
+          uniformTypeIdentifiers: <String>['public.image'],
+          webWildCards: <String>['image/*']);
+
+      await plugin
+          .getSaveLocation(acceptedTypeGroups: <XTypeGroup>[group, groupTwo]);
+
+      final VerificationResult result =
+          verify(mockApi.displaySavePanel(captureAny));
+      final SavePanelOptions options = result.captured[0] as SavePanelOptions;
+      expect(options.allowedFileTypes!.extensions, <String>['txt', 'jpg']);
+      expect(options.allowedFileTypes!.mimeTypes,
+          <String>['text/plain', 'image/jpg']);
+      expect(options.allowedFileTypes!.utis,
+          <String>['public.text', 'public.image']);
+    });
+
+    test('passes initialDirectory correctly', () async {
+      await plugin.getSaveLocation(
+          options:
+              const SaveDialogOptions(initialDirectory: '/example/directory'));
+
+      final VerificationResult result =
+          verify(mockApi.displaySavePanel(captureAny));
+      final SavePanelOptions options = result.captured[0] as SavePanelOptions;
+      expect(options.directoryPath, '/example/directory');
+    });
+
+    test('passes confirmButtonText correctly', () async {
+      await plugin.getSaveLocation(
+          options: const SaveDialogOptions(confirmButtonText: 'Open File'));
+
+      final VerificationResult result =
+          verify(mockApi.displaySavePanel(captureAny));
+      final SavePanelOptions options = result.captured[0] as SavePanelOptions;
+      expect(options.prompt, 'Open File');
+    });
+
+    test('throws for a type group that does not support macOS', () async {
+      const XTypeGroup group = XTypeGroup(
+        label: 'images',
+        webWildCards: <String>['images/*'],
+      );
+
+      await expectLater(
+          plugin.getSaveLocation(acceptedTypeGroups: <XTypeGroup>[group]),
+          throwsArgumentError);
+    });
+
+    test('allows a wildcard group', () async {
+      const XTypeGroup group = XTypeGroup(
+        label: 'text',
+      );
+
+      await expectLater(
+          plugin.getSaveLocation(acceptedTypeGroups: <XTypeGroup>[group]),
+          completes);
+    });
+
+    test('ignores all type groups if any of them is a wildcard', () async {
```

Comment:
Some of these tests are not specific of `getSaveLocation`, but seem to be indirectly testing `_allowedTypesFromTypeGroups`.

Should `_allowedTypesFromTypeGroups` be made `@visibleForTesting` and have the tests thrown at it directly?

Reply:
They have to do with `getSaveLocation` though, because this is the behavior `getSaveLocation` needs to have. It's true that it indirectly tests `_allowedTypesFromTypeGroups`, but indirectly testing internal implementation details (like that something happens to have been extracted into a private helper) is a feature, not a bug. Exposing internals for testing should only be a last resort when something can't be tested via public API, because it makes the tests more fragile and thus internal refactoring harder (and more error prone since you have to change tests at the same time as the implementation is being changed).